File: resources/chats.go
package resources

import (
	"context"
	"time"
)

// Chats handles chat-related API operations
type Chats struct {
	client    ClientInterface
	Reminders *Reminders
}

// NewChats creates a new Chats resource client
func NewChats(client ClientInterface) *Chats {
	return &Chats{
		client:    client,
		Reminders: NewReminders(client),
	}
}

// Chat represents a chat/conversation
type Chat struct {
	ID                     string           `json:"id"`
	AccountID              string           `json:"accountID"`
	Network                string           `json:"network"`
	Title                  string           `json:"title"`
	Type                   string           `json:"type"` // single, group
	UnreadCount            int              `json:"unreadCount"`
	Participants           ChatParticipants `json:"participants"`
	IsArchived             *bool            `json:"isArchived,omitempty"`
	IsMuted                *bool            `json:"isMuted,omitempty"`
	IsPinned               *bool            `json:"isPinned,omitempty"`
	LastActivity           *string          `json:"lastActivity,omitempty"`
	LastReadMessageSortKey interface{}      `json:"lastReadMessageSortKey,omitempty"` // string or number
	LocalChatID            *string          `json:"localChatID,omitempty"`
}

// ChatParticipants represents chat participants information
type ChatParticipants struct {
	HasMore bool   `json:"hasMore"`
	Items   []User `json:"items"`
	Total   int    `json:"total"`
}

// ChatCreateParams represents parameters for creating a chat
type ChatCreateParams struct {
	AccountID      string   `json:"accountID"`
	ParticipantIDs []string `json:"participantIDs"`
	Type           string   `json:"type"` // single, group
	Title          *string  `json:"title,omitempty"`
}

// ChatCreateResponse represents the response from creating a chat
type ChatCreateResponse struct {
	Chat    Chat   `json:"chat"`
	Success bool   `json:"success"`
	Error   string `json:"error,omitempty"`
}

// ChatRetrieveParams represents parameters for retrieving a chat
type ChatRetrieveParams struct {
	ChatID string `json:"chatID"`
}

// ChatArchiveParams represents parameters for archiving a chat
type ChatArchiveParams struct {
	ChatID   string `json:"chatID"`
	Archived bool   `json:"archived"`
}

// ChatSearchParams represents parameters for searching chats
type ChatSearchParams struct {
	AccountIDs   []string `json:"accountIDs,omitempty"`
	ChatType     *string  `json:"chatType,omitempty"`
	IncludeMuted *bool    `json:"includeMuted,omitempty"`
	Limit        *int     `json:"limit,omitempty"`
	Cursor       *string  `json:"cursor,omitempty"`
	Scope        *string  `json:"scope,omitempty"`
	Query        *string  `json:"query,omitempty"`
}

// ChatsCursor represents paginated chat results
type ChatsCursor = Cursor[Chat]

// Create creates a single or group chat on a specific account
func (c *Chats) Create(ctx context.Context, params ChatCreateParams) (*ChatCreateResponse, error) {
	var result ChatCreateResponse
	err := c.client.DoRequest(ctx, "POST", "/v0/create-chat", params, &result)
	if err != nil {
		return nil, err
	}
	return &result, nil
}

// Retrieve gets chat details including metadata, participants, and latest message
func (c *Chats) Retrieve(ctx context.Context, params ChatRetrieveParams) (*Chat, error) {
	var result Chat
	err := c.client.DoRequest(ctx, "GET", "/v0/get-chat", params, &result)
	if err != nil {
		return nil, err
	}
	return &result, nil
}

// Archive archives or unarchives a chat
func (c *Chats) Archive(ctx context.Context, params ChatArchiveParams) (*BaseResponse, error) {
	var result BaseResponse
	err := c.client.DoRequest(ctx, "POST", "/v0/archive-chat", params, &result)
	if err != nil {
		return nil, err
	}
	return &result, nil
}

// Search searches chats by title/network or participants
func (c *Chats) Search(ctx context.Context, params ChatSearchParams) (*ChatsCursor, error) {
	var result ChatsCursor
	err := c.client.DoRequest(ctx, "GET", "/v0/search-chats", params, &result)
	if err != nil {
		return nil, err
	}
	return &result, nil
}

// Reminders handles chat reminder operations
type Reminders struct {
	client ClientInterface
}

// NewReminders creates a new Reminders resource client
func NewReminders(client ClientInterface) *Reminders {
	return &Reminders{client: client}
}

// ReminderCreateParams represents parameters for creating a reminder
type ReminderCreateParams struct {
	ChatID    string    `json:"chatID"`
	Timestamp time.Time `json:"timestamp"`
	Message   *string   `json:"message,omitempty"`
}

// ReminderDeleteParams represents parameters for deleting a reminder
type ReminderDeleteParams struct {
	ChatID string `json:"chatID"`
}

// Create sets a reminder for a chat at a specific time
func (r *Reminders) Create(ctx context.Context, params ReminderCreateParams) (*BaseResponse, error) {
	var result BaseResponse
	err := r.client.DoRequest(ctx, "POST", "/v0/set-chat-reminder", params, &result)
	if err != nil {
		return nil, err
	}
	return &result, nil
}

// Delete clears a chat reminder
func (r *Reminders) Delete(ctx context.Context, params ReminderDeleteParams) (*BaseResponse, error) {
	var result BaseResponse
	err := r.client.DoRequest(ctx, "POST", "/v0/clear-chat-reminder", params, &result)
	if err != nil {
		return nil, err
	}
	return &result, nil
}

File: beeper-rust-sdk/src/resources/chats.rs
use crate::client::BeeperDesktop;
use crate::error::Result;
use crate::resources::shared::{Chat, ChatsCursor, BaseResponse};
use chrono::{DateTime, Utc};
use reqwest::Method;
use serde::{Deserialize, Serialize};

/// Chats handles chat-related API operations
#[derive(Debug, Clone)]
pub struct Chats {
    client: BeeperDesktop,
    pub reminders: Reminders,
}

impl Chats {
    /// Create a new Chats resource client
    pub fn new(client: BeeperDesktop) -> Self {
        let reminders = Reminders::new(client.clone());
        Self { client, reminders }
    }

    /// Create creates a single or group chat on a specific account
    pub async fn create(&self, params: &ChatCreateParams) -> Result<ChatCreateResponse> {
        self.client
            .do_request(Method::POST, "/v0/create-chat", Some(params))
            .await
    }

    /// Retrieve gets chat details including metadata, participants, and latest message
    pub async fn retrieve(&self, params: &ChatRetrieveParams) -> Result<Chat> {
        self.client
            .do_request(Method::GET, "/v0/get-chat", Some(params))
            .await
    }

    /// Archive archives or unarchives a chat
    pub async fn archive(&self, params: &ChatArchiveParams) -> Result<BaseResponse> {
        self.client
            .do_request(Method::POST, "/v0/archive-chat", Some(params))
            .await
    }

    /// Search searches chats by title/network or participants
    pub async fn search(&self, params: &ChatSearchParams) -> Result<ChatsCursor> {
        self.client
            .do_request(Method::GET, "/v0/search-chats", Some(params))
            .await
    }
}

/// Reminders handles chat reminder operations
#[derive(Debug, Clone)]
pub struct Reminders {
    client: BeeperDesktop,
}

impl Reminders {
    /// Create a new Reminders resource client
    pub fn new(client: BeeperDesktop) -> Self {
        Self { client }
    }

    /// Create sets a reminder for a chat at a specific time
    pub async fn create(&self, params: &ReminderCreateParams) -> Result<BaseResponse> {
        self.client
            .do_request(Method::POST, "/v0/set-chat-reminder", Some(params))
            .await
    }

    /// Delete clears a chat reminder
    pub async fn delete(&self, params: &ReminderDeleteParams) -> Result<BaseResponse> {
        self.client
            .do_request(Method::POST, "/v0/clear-chat-reminder", Some(params))
            .await
    }
}

/// ChatCreateParams represents parameters for creating a chat
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChatCreateParams {
    pub account_id: String,
    pub participant_ids: Vec<String>,
    /// Type of chat (single, group)
    #[serde(rename = "type")]
    pub chat_type: String,
    pub title: Option<String>,
}

/// ChatCreateResponse represents the response from creating a chat
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatCreateResponse {
    pub chat: Chat,
    pub success: bool,
    pub error: Option<String>,
}

/// ChatRetrieveParams represents parameters for retrieving a chat
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChatRetrieveParams {
    pub chat_id: String,
}

/// ChatArchiveParams represents parameters for archiving a chat
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChatArchiveParams {
    pub chat_id: String,
    pub archived: bool,
}

/// ChatSearchParams represents parameters for searching chats
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChatSearchParams {
    pub account_ids: Option<Vec<String>>,
    pub chat_type: Option<String>,
    pub include_muted: Option<bool>,
    pub limit: Option<i32>,
    pub cursor: Option<String>,
    pub scope: Option<String>,
    pub query: Option<String>,
}

impl ChatSearchParams {
    /// Create a new ChatSearchParams with default values
    pub fn new() -> Self {
        Self::default()
    }
}

/// ReminderCreateParams represents parameters for creating a reminder
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReminderCreateParams {
    pub chat_id: String,
    pub timestamp: DateTime<Utc>,
    pub message: Option<String>,
}

/// ReminderDeleteParams represents parameters for deleting a reminder
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReminderDeleteParams {
    pub chat_id: String,
}